### ISSUE TITLE
ks_mysql not found on mac OS X mariadb - brew install

### ISSUE THREAD

**Issue by XavRsl** [willbryant/kitchen_sync#46]
Hi!
I'm trying to copy a DB from a server to my local machine. I get the error:
```Couldn't exec ks_mysql: No such file or directory```

I installed KS using brew and I didn't install MySQL locally but MariaDB. Unfortunately, I can't install MySQL as well as MariaDB using brew (Error: Cannot install MySQL because conflicting formulae are installed. MySQL, MariaDB, and Percona install the same binaries). 

Is there any way I could create (build) ks_mysql using MariaDB?

Thanks!

**Comment by XavRsl** [comment]
Here you go:

Brew list:
```
apr			freetype		isl			libpq			mhash			parallel		supervisor
apr-util		fribidi			jasper			librsvg			mpfr			pcre			swig
argon2			gcc			jpeg			libsodium		mysql-client		php@7.1			telnet
aspell			gdbm			jpegoptim		libtiff			nginx			pixman			tor
autoconf		gdk-pixbuf		jq			libtool			node@8			pkg-config		unixodbc
beanstalkd		gettext			kibana			libunistring		nspr			pngquant		vips
boost			giflib			kitchen-sync		libxml2			nss			poppler			watch
cairo			gifsicle		libcroco		libyaml			oniguruma		python			webp
cscope			glib			libevent		libzip			openjpeg		readline		wget
dnsmasq			gmp			libexif			little-cms2		openldap		redis			xz
docbook			go			libffi			macvim			openssl			rename			yaml-cpp
docbook-xsl		graphicsmagick		libgsf			mailhog			openssl@1.1		ruby
elasticsearch		graphite2		libidn2			mariadb			optipng			source-highlight
fftw			harfbuzz		libmemcached		mcrypt			orc			sphinx-doc
fontconfig		icu4c			libmpc			memcache-top		packer			sqlite
freetds			imagemagick		libpng			memcached		pango			ssh-copy-id
```
Formulae install output

```
==> Installing kitchen-sync
==> Downloading https://homebrew.bintray.com/bottles/kitchen-sync-1.5.mojave.bottle.tar.gz
######################################################################## 100.0%
==> Pouring kitchen-sync-1.5.mojave.bottle.tar.gz
🍺  /usr/local/Cellar/kitchen-sync/1.5: 5 files, 77.4KB
==> Caveats
==> openssl
A CA file has been bootstrapped using certificates from the SystemRoots
keychain. To add additional certificates (e.g. the certificates added in
the System keychain), place .pem files in
  /usr/local/etc/openssl/certs

and run
  /usr/local/opt/openssl/bin/c_rehash

openssl is keg-only, which means it was not symlinked into /usr/local,
because Apple has deprecated use of OpenSSL in favor of its own TLS and crypto libraries.

If you need to have openssl first in your PATH run:
  echo 'export PATH="/usr/local/opt/openssl/bin:$PATH"' >> ~/.zshrc

For compilers to find openssl you may need to set:
  export LDFLAGS="-L/usr/local/opt/openssl/lib"
  export CPPFLAGS="-I/usr/local/opt/openssl/include"

For pkg-config to find openssl you may need to set:
  export PKG_CONFIG_PATH="/usr/local/opt/openssl/lib/pkgconfig"

==> mysql-client
mysql-client is keg-only, which means it was not symlinked into /usr/local,
because conflicts with mysql.

If you need to have mysql-client first in your PATH run:
  echo 'export PATH="/usr/local/opt/mysql-client/bin:$PATH"' >> ~/.zshrc

For compilers to find mysql-client you may need to set:
  export LDFLAGS="-L/usr/local/opt/mysql-client/lib"
  export CPPFLAGS="-I/usr/local/opt/mysql-client/include"

For pkg-config to find mysql-client you may need to set:
  export PKG_CONFIG_PATH="/usr/local/opt/mysql-client/lib/pkgconfig"
```

**Comment by willbryant** [comment]
Hmm yeah, this rings a bell, but I can't remember what the resolution was (I don't maintain the recipe).  I think I was told that the root of the problem is that the homebrew recipe just says "mysql", not "mysql or mariadb" - not sure if there is a way to do that.

I think you can use "brew install kitchen-sync --without-mysql-client" to stop it from installing the mysql version of the library, and then you'd just need to help it find the mariadb copy.

But, I believe you probably don't actually have the mariadb client library headers, because the only recipe I see in your list is mariadb, and that will just have the binaries.

Could you check that - do you have a copy of mysql.h anywhere under /usr/local/Cellar/mariadb?

**Comment by willbryant** [comment]
Whoa OK, I just looked at the bottled version of kitchen sync, and it is missing the actual binaries that do the work.  So I guess the recipe is a bit broken at the moment.

It doesn't even work when I try `brew install kitchen-sync --build-from-source --without-mysql-client`, so I think something's gone wrong with the recipe at some point.  The weird thing is that the recipe uses the same build steps that I use for my own source builds, so I guess the problem must be its environment setup.

If you're in a hurry to get KS up and running, please follow the "To compile from source yourself" steps in https://github.com/willbryant/kitchen_sync/blob/master/INSTALL.md.

**Comment by XavRsl** [comment]
Yes, 
```
$ locate mysql.h
/usr/local/Cellar/mariadb/10.1.20/include/mysql/mysql.h
/usr/local/Cellar/mariadb/10.1.20/include/mysql/private/probes_mysql.h
/usr/local/Cellar/mariadb/10.2.12/include/mysql/mysql.h
/usr/local/Cellar/mariadb/10.2.12/include/mysql/server/mysql.h
/usr/local/Cellar/mariadb/10.2.12/include/mysql/server/private/probes_mysql.h
/usr/local/Cellar/mysql-client/5.7.23/include/mysql/mysql.h

$ cd /usr/local/include/mysql
$ pwd -P
/usr/local/Cellar/mariadb/10.2.12/include/mysql
```

so it seems to match with the CMAKE
```
FIND_PATH(MYSQL_INCLUDE_DIR mysql.h
  /usr/local/include/mysql
  /usr/include/mysql
  /usr/include/mariadb
  /usr/local/include/mariadb
)
```

I'll compile it from source myself but if you need me to reinstall it from brew later for testing purposes, let me know.

Thanks a lot for your time.

Xavier

**Comment by XavRsl** [comment]
Just so you know, building from source worked on OSX. Thanks.

I've also installed it on CentOS and here is what could be added to the docs:

CentOS 7
----------

You can install the above build dependencies on CentOS 7 using:
```
yum install gcc gcc-c++ make cmake openssl boost-devel yaml-cpp-devel openssl-devel
```

And one or both of:
```
yum install postgresql-devel
yum install mariadb-devel
```

MariaDB is now the default replacement for mysql on CentOS.

To build, change to the kitchen_sync directory where you checked out the files, then:
```
  cd build
  cmake .. && make && make install
```


Do you want me to make a PR for it?

Thanks

**Comment by willbryant** [comment]
Looks like it's something to do with the `--env=std` option to clang, which is new to me, so I'll have to do some digging.

Good to hear you got it installed now.  Thanks very much for putting together the CentOS notes, please do go ahead and send that through as a PR.  I guess put it between Ubuntu and OS X?

**Comment by willbryant** [comment]
I've released 1.6 which has a couple of improvements here:
* better error messages explaining what 'No such file or directory' implies here
* build-time warning when no database clients are find
* a lot of extra content in INSTALL.md explaining how this all works

In 1.7 that build-time warning will change to be fatal (by default), which should stop Homebrew producing a useless build - I decided to push this back to the next release so that the better error messages can be released first.

But we still need to actually fix the recipe :).

**Comment by willbryant** [comment]
OK well that didn't really reveal the problem, at least in my local build attempts (`brew install --build-from-source --keep-tmp --HEAD kitchen-sync`).  The CMake step works fine and *does* find mysql/mariadb with no warning, but the compilation fails with `fatal error: 'mysql.h' file not found`.

**Comment by willbryant** [comment]
This was pretty brutal.  I had to monkey-patch Homebrew to find out what it was really doing as it was playing games using compiler wrapper scripts and they override the compiler include paths!

Please `brew edit kitchen-sync` and try the following revised version:

```ruby
class KitchenSync < Formula
  desc "Fast efficiently sync database without dumping & reloading"
  homepage "https://github.com/willbryant/kitchen_sync"
  url "https://github.com/willbryant/kitchen_sync/archive/1.7.tar.gz"
  sha256 "c8f393056290de5488d46b975a574638bb33c2d1f697d4902f0d46f662112f86"
  head "https://github.com/willbryant/kitchen_sync.git"

  deprecated_option "without-mysql" => "without-mysql-client"

  depends_on "cmake" => :build
  depends_on "libpq"
  depends_on "mysql-client"

  needs :cxx14

  def install
    # If the mysql or mariadb formulae are installed, FindMySQL.cmake will find their mysql.h at the symlinked location
    # /usr/local/include/mysql, but unfortunately then Homebrew's clang++ shim will remove the -I/usr/local/include/mysql
    # option from the command line.  Tell FindMySQL to find the mysql-client copy specifically.  Do the same for libpq.
    system "cmake", ".", *std_cmake_args,
      "-DMySQL_INCLUDE_DIR=#{Formula["mysql-client"].opt_include}/mysql",
      "-DMySQL_LIBRARY_DIR=#{Formula["mysql-client"].opt_lib}",
      "-DPostgreSQL_INCLUDE_DIR=#{Formula["libpq"].opt_include}",
      "-DPostgreSQL_LIBRARY_DIR=#{Formula["libpq"].opt_lib}"
    system "make", "install"
  end

  test do
    output = shell_output("#{bin}/ks --from a://b/ --to c://d/ 2>&1")
    assert_match "Finished Kitchen Syncing", output
  end
end
```

**Comment by willbryant** [comment]
The critical bit of that change (for mysql) was fixed in Homebrew commit f6a72beb2831b06a3a03bf7d2e5396978afafddc by Igor Kapkov, and I've submitted a separate PR for the libpq inclusion.
